build.gradle.kts

```diff
@@ -6,43 +6,82 @@ import org.gradle.util.GradleVersion
 import java.nio.file.Files
 import java.nio.file.StandardCopyOption
 import java.text.SimpleDateFormat
-import java.util.*
+import java.util.Date
 
 plugins {
     java
     signing
     `maven-publish`
     idea
-    id("com.adarshr.test-logger") version "2.1.1"
     id("com.diffplug.gradle.spotless") version "3.25.0"
     id("com.gradle.plugin-publish") version "0.12.0"
-    id("io.freefair.lombok") version "4.1.6" apply false
     id("net.linguica.maven-settings") version "0.5"
     id("net.researchgate.release") version "2.8.1"
     id("org.ajoberstar.grgit") version "4.1.0"
-    id("org.kordamp.gradle.jacoco") version "0.42.1"
 
-    if ( org.gradle.util.GradleVersion.current().compareTo(org.gradle.util.GradleVersion.version("6.0")) < 0) {
-        id("com.github.johnrengelman.shadow") version "5.2.0"
+    when {
+        org.gradle.util.GradleVersion.current() < org.gradle.util.GradleVersion.version("5.0") -> {
+            id("com.adarshr.test-logger") version "1.7.1"
+            // XXX: Versions 4.x > 4.0.1 suffer from <https://github.com/johnrengelman/shadow/issues/425>
+            id("com.github.johnrengelman.shadow") version "4.0.1"
+        }
+        org.gradle.util.GradleVersion.current() < org.gradle.util.GradleVersion.version("6.0") -> {
+            id("com.adarshr.test-logger") version "2.1.1"
+            id("com.github.johnrengelman.shadow") version "5.2.0"
+        }
+        org.gradle.util.GradleVersion.current() < org.gradle.util.GradleVersion.version("7.0") -> {
+            id("com.adarshr.test-logger") version "2.1.1"
+            id("com.github.johnrengelman.shadow") version "6.1.0"
+        }
+        else -> {
+            id("com.adarshr.test-logger") version "3.0.0"
+            id("com.github.johnrengelman.shadow") version "7.0.0"
+        }
+    }
 
+    if (org.gradle.util.GradleVersion.current() < org.gradle.util.GradleVersion.version("5.0")) {
+        id("io.freefair.lombok") version "2.9.5" apply false
+    } else if (org.gradle.util.GradleVersion.current() < org.gradle.util.GradleVersion.version("5.2")) { 
+        // XXX: Versions 3.x < 3.6.1 suffer from <https://github.com/freefair/gradle-plugins/issues/31>
+        id("io.freefair.lombok") version "3.0.0" apply false
+    } else if (org.gradle.util.GradleVersion.current() < org.gradle.util.GradleVersion.version("6.0")) {
+        id("io.freefair.lombok") version "4.1.6" apply false
     } else {
-        id("com.github.johnrengelman.shadow") version "6.1.0"
+        id("io.freefair.lombok") version "5.3.3.3" apply false
+    }
+
+    if (org.gradle.util.GradleVersion.current() >= org.gradle.util.GradleVersion.version("5.3")) {
+        id("org.kordamp.gradle.jacoco") version "0.46.0"
     }
 }
 
-apply(plugin = "net.researchgate.release")
+if (org.gradle.util.GradleVersion.current() < org.gradle.util.GradleVersion.version("5.3")) {
+    tasks.register("AggregateJacocoReport");
```

Comment:
Neat! Wonder what happens if there is no report when codecov task runs in GitHub? .... I presume it just ignores it.

Also, could we add a comment line as to why as I think this is something that isn't obvious at first glance

Reply:
I wonder how it handles multiple reports in general, since we want it to use the one with the highest coverage.

Reply:
Yeah I was wondering that - hehe originally thinking of running coverage for only one version. Unless it accumulates them?

Reply:
I thought it aggregates them. At the very least, you can sometimes see the results change before your eyes if you watch a PR where the runs are in progress.

Reply:
Yeah I think it does merge them. 
Can we get a comment line to explain why an empty task has been added to resolve this?

Reply:
Added comment.